### ISSUE TITLE
[RFC] TensorRT Model Optimizer - Product Roadmap

### ISSUE THREAD

**Issue by hchings** [NVIDIA/TensorRT-Model-Optimizer#108]
# TensorRT Model Optimizer - Product Roadmap
[TensorRT Model Optimizer's](https://github.com/NVIDIA/TensorRT-Model-Optimizer) goal is to provide a unified API library that enables our developers to easily achieve state of the art model optimizations resulting in the best inference speed-ups. Model Optimizer will continuously enhance its existing features leveraging advanced capabilities to introduce new cutting-edge techniques and stay at the forefront of AI model optimization.

In striving for this, our roadmap and development follow these product strategies:

1. Provide a one-stop-shop for SOTA optimization methods (quantization, distillation, sparsity, pruning, speculation, etc) with easy-to-use APIs for developers to chain different methods with reproducibility.
2. Provide transparency and extensibility, making it easy for developers and researchers to innovate and contribute.
3. Provide the best easy-to-use recipes in the ecosystem through software-hardware co-design on NVIDIA platforms. Since Model-Optimizer's launch, we’ve been delivering 50% to ~5x speedup on top of existing runtime and compiler optimizations on NVIDIA GPUs with minimal impact on model accuracy ([Latest News](https://github.com/NVIDIA/TensorRT-Model-Optimizer?tab=readme-ov-file#latest-news)).
4. Tightly integrate into the Deep Learning inference and training ecosystem, beyond NVIDIA’s in-house stacks. Offer many-to-many optimizations by supporting popular frameworks like [vLLM](https://github.com/vllm-project/vllm), [SGLang](https://github.com/sgl-project/sglang), [TensorRT-LLM](https://github.com/NVIDIA/TensorRT-LLM), [TensorRT](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#overview), [NeMo](https://github.com/NVIDIA/NeMo), and [Hugging Face](https://huggingface.co/).

In the following sections, we outline our key investment areas and upcoming features. All are subject to change and we’ll update this doc regularly. Our goal of sharing roadmaps is to increase visibility of Model-Optimizer's directions and upcoming features. We welcome any questions and feedback in this thread and feature requests in github Issues 😊.

### 1. FP4 inference on NVIDIA Blackwell
- Optimized FP4 [Post Training Quantization (PTQ)](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_ptq#post-training-quantization-ptq).
- Optimized FP4 [Quantization Aware Training (QAT)](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_qat#quantization-aware-training-qat).
- Advanced quantization techniques (#model-optimization-techniques).
### 2. Model optimization techniques
- Advanced PTQ methods, QAT with distillation, Attention Sparsity, token-efficient pruning/distillation, general rotation/smoothing.
- Easier E2E Speculation/Distillation/QAT.
- Improve caching techniques.
- Host pre-optimized checkpoints on HuggingFace.
### 3. Developer Productivity
- Open sourced for everyone with improved extensibility, transparency, debuggability and accessibility.
- Docs/code improvements to assist community to expand on their use cases, experiment and introduce custom optimizations.
- Ready-to-deploy optimized checkpoints for both ease of use and resource limited developers.
### 4. Choice of Deployment
- Support for TRT-LLM, vLLM and SGLang expanding.
- In-framework deployment for quick prototyping and deploying models not yet supported in TRT-LLM/vLLM/SGLang.
### 5. Expand Support Matrix
- Expand data-type availability.
- Continuously expand our model support based on community interests.
- Continued Auto and Windows support

### 1. FP4 inference on NVIDIA Blackwell
[NVIDIA Blackwell platform](https://www.nvidia.com/en-us/data-center/technologies/blackwell-architecture/) powers a new era of computing with FP4 AI inference capabilities. Model-Optimizer has provided initial FP4 recipes and quantization techniques and will continue to improve FP4 with advanced techniques:
1. For the majority of developers, Model-Optimizer offers **[Post Training Quantization (PTQ)](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_ptq#post-training-quantization-ptq)** (weight and activation, weight-only) and our proprietary AutoQuantize for FP4 inference. [AutoQuantize](https://nvidia.github.io/TensorRT-Model-Optimizer/guides/_pytorch_quantization.html#optimal-partial-quantization-using-autoquantize-auto-quantize) automates per-layer quantization formats to achieve minimal model accuracy loss.
2. For developers who require lossless FP4 quantization, Model-Optimizer offers **[Quantization Aware Training (QAT)](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_qat#quantization-aware-training-qat)**, which makes the neural network more resilient to quantization. Model-Optimizer QAT already works with [NVIDIA Megatron](https://github.com/NVIDIA/Megatron-LM), [NVIDIA NeMo](https://www.nvidia.com/en-us/ai-data-science/products/nemo/), native PyTorch training, and [Hugging Face Trainer](https://huggingface.co/docs/transformers/en/main_classes/trainer).

#### 2.1 Model compression algorithms
Model-Optimizer collaborates with Nvidia and external research labs to continuously develop and integrate state-of-the-art techniques into our library for faster inference. Our recent focus areas include:
- Advanced PTQ methods (e.g., [SVDQuant](https://svdquant.mit.edu/), [QuaRot](https://arxiv.org/abs/2404.00456), [SpinQuant](https://arxiv.org/abs/2405.16406))
- QAT with distillation, a proven path for FP4 inference
- Attention sparsity (e.g., [SnapKV](https://arxiv.org/abs/2404.14469), [DuoAttention](https://arxiv.org/abs/2410.10819))
- [AutoQuantize](https://nvidia.github.io/TensorRT-Model-Optimizer/guides/_pytorch_quantization.html#optimal-partial-quantization-using-autoquantize-auto-quantize) improvements (e.g., support more fine-trained format selection and various weight and activation combination)
- New token-efficient pruning and distillation methods
- Infrastructure to support general rotation and smoothing

#### 2.2 Optimized techniques for LLM and VLM
Model-Optimizer works with [TensorRT-LLM](https://github.com/NVIDIA/TensorRT-LLM), [vLLM](https://github.com/vllm-project/vllm) and [SGLang](https://github.com/sgl-project/sglang) to streamline optimized model deployment. This includes expanding focus on model optimizations that require finetuning. To allow streamlined experience, Model-optimizer is working with (Hugging Face/ NVIDIA [NeMo](https://www.nvidia.com/en-us/ai-data-science/products/nemo/) and [Megatron-LM](https://github.com/NVIDIA/Megatron-LM)) to deliver exceptional E2E solution for these optimizations. Our focus areas include:
- (Speculation) Integrated draft model: Medusa, Redrafter, MTP and [EAGLE](https://arxiv.org/abs/2401.15077).
- (Speculation/Distillation) Standalone draft model training through pruning and knowledge distillation. 
- (Distillation) Standalone model shrinking/compressing through pruning and knowledge distillation (e.g. Llama-3.2 1/3B).
- (Quantization) Quantization aware training with support of FP8 and FP4.
- Out-of-the-box deployment with trtllm-serve, [NVIDIA NIM](https://www.nvidia.com/en-us/ai/), and vLLM serve.
- Hosting pre-optimized checkpoints for popular models such as DeepSeek-R1, Llama-3.1, Llama-3.3 and Nemotron family on [Hugging Face Model-Optimizer collection](https://huggingface.co/collections/nvidia/model-optimizer-66aa84f7966b3150262481a4).

#### 2.3 Optimized techniques for diffusers
Model-Optimizer will continue to accelerate image generation inference by investing in these areas:
- Quantization: Expand model support for INT8/FP8/FP4 PTQ and QAT. e.g., [FLUX](https://github.com/black-forest-labs/flux) model series.
- Caching: Adding more training-free and lightweight finetuning-based caching techniques with user-friendly APIs. (Previous work: [Cache Diffusion](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/diffusers/cache_diffusion)).
- Improve easy of use of the deployment pipelines, including adding multi-GPU support.

#### 3.1 Open-sourcing
To provide extensibility and transparency for everyone, Model-Optimizer is now Open Source! Paired with continued documentation/code additions to improve extensibility/usability, Model-Optimizer will continue to have a large focus on enabling our community to expand and contribute for their own use-cases. This will enable developers, for example, to experiment with custom calibration algorithms or contribute to the latest techniques. Users can also self-service to add model support or non-standard data-types, and benefit from improved debuggability and accessibility. 

#### 3.2 Ready-to-deploy optimized checkpoints
For developers who have limited GPU resources to optimize large models or prefer to skip the optimization steps, we currently offer quantized checkpoints of popular models in the [Hugging Face Model Optimizer collection](https://huggingface.co/collections/nvidia/model-optimizer-66aa84f7966b3150262481a4). Developers can deploy these optimized checkpoints directly on TensorRT-LLM, vLLM and SGLang (Depending on the checkpoint). We currently have published FP8/FP4/Medusa Llama family model checkpoints and FP4 checkpoint for DeepSeek-R1. In the near future we are working to expand to optimized FLUX, diffusion, Medusa-trained checkpoints, Eagle-trained checkpoints and more.

#### 4.1 Popular Community Frameworks
To offer greater flexibility, we’ve been investing in supporting popular inference and serving frameworks like [vLLM](https://github.com/vllm-project/vllm) and [SGLang](https://github.com/sgl-project/sglang), in addition to having seamless integration with the NVIDIA AI software ecosystem. We currently provide an [initial workflow for vLLM deployment](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/examples/llm_ptq/README.md#deploy-fp8-quantized-model-using-vllm) and an example for deploying [Unified HuggingFace Checkpoint](https://nvidia.github.io/TensorRT-Model-Optimizer/deployment/3_unified_hf.html), with more model support planned.

#### 4.2 In-Framework Deployment
We have enabled and released a path for deployment within native PyTorch. This decouples model build/compile from runtime and offers several benefits:
1. When optimizing inference performance or exploring new model compression techniques, Model-Optimizer users can quickly prototype in the PyTorch runtime and native PyTorch APIs to evaluate performance gains. Once satisfied, they can transition to the TensorRT-LLM runtime as the final step to maximize performance.
2. For models not yet supported by TensorRT-LLM or applications that do not need ultra-fast inference speeds, users can get out-of-the-box performance improvements within native PyTorch.

Developers can utilize [AutoDeploy](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_autodeploy#deploy-autoquant-models-with-autodeploy) or [Real Quantization](https://github.com/NVIDIA/TensorRT-Model-Optimizer/tree/main/examples/llm_qat#end-to-end-qlora-with-real-quantization) for these in-framework deployments.

#### 5.1 Data types
Alongside our existing [supported dtypes](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/examples/llm_ptq/README.md#model-support-list), we’ve recently added [MXFP4](https://arxiv.org/pdf/2310.10537) support and will soon expand to emerging popular dtypes like [FP6](https://github.com/microsoft/DeepSpeed/tree/master/blogs/deepspeed-fp6/03-05-2024) and sub-4-bit. Our focus is to further speed up GenAI inference with the least possible impact on model fidelity.

#### 5.2 Model Support
We strive to streamline our techniques to provide the shortest time from new model/feature to optimized model. This provides our community with the shortest time to deploy. We’ll continue to expand LLM/Diffusion model support, invest more in LLM with multi-modality (vision, video, audio, image generation, and action), and continuously expand our model support based on community interests.

#### 5.3 Platform & Other Support
Model-Optimizer's explicit quantization will be part of the upcoming [NVIDIA DriveOS](https://developer.nvidia.com/drive/os) releases. We recently added an e2e [BEVFormer INT8 example](https://github.com/NVIDIA/DL4AGX/tree/master/AV-Solutions) in NVIDIA DL4AGX, with more model support coming soon for Automotive customers. Model-Optimizer also has planned support for ONNX FP4 for [DRIVE Thor](https://nvidianews.nvidia.com/news/nvidia-unveils-drive-thor-centralized-car-computer-unifying-cluster-infotainment-automated-driving-and-parking-in-a-single-cost-saving-system).

In Q4 2024, Model-Optimizer added formal support for Windows (see [Model-Optimizer-Windows](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/examples/windows/README.md)), targeting Windows RTX PC systems with tight integration with Windows ecosystem such as [torch.onnx.export](https://pytorch.org/docs/stable/onnx.html), [HuggingFace-Optimum](https://huggingface.co/docs/optimum/en/exporters/onnx/usage_guides/export_a_model/), [GenAI](https://github.com/microsoft/onnxruntime-genai/), and [Olive](https://github.com/microsoft/Olive/). It currently supports quantization such as INT4 AWQ, INT8, FP8 and we’ll expand to more techniques suitable for Windows.

## Upcoming releases
We'll do our best to provide visibility into our upcoming releases. Details are subject to change and this table is not comprehensive.

| | Model-Optimizer v0.23 (Jan 2025) | Model-Optimizer v0.25 (Feb 2025) | Model-Optimizer v0.27 (Mar 2025) |
|------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------:|
| Blackwell Inference | PTQ, QAT recipes for FP4 Inference | FP4 AWQ Support | - PTQ FP4/FP8 W4A8 recipe <br/> - PTQ FP4 ONNX <br/> - PTQ FP8 Blockwise |
| Feature Improvements | Added Minitron pruning support for NeMo 2.0 GPT models | - MXFP8/MXFP4 eval example <br/> - Advanced FP8 KV cache quantization​​ <br/> - Add rotation-based quantization support for methods like QuaRot <br/> - FSDP2 QAT support | - Advanced FP4 KV cache quantization​​ <br/> - SVDQuant |
| Developer Productivity | - Model Optimizer is now OSS! <br/> - FP4 quantization examples | - Pruning example for NeMo 2.0 <br/> - Optimized quantization simulation support | MXFP8 ONNX support example |
| Model Support | - Llama 3.3 <br/> - Phi 4 | - DeepSeek-R1 FP4 <br/> - Llama3.3-70B FP4 <br/> - Llama3.1-405B FP4 | DeepSeek-R1 Spec Decoding |
| Platform Support & Ecosystem | Unified HF checkpoint llama3.1 FP8 with SGLang | FP8 KV Cache vLLM/SGLang export | INT8/FP8 support for Windows |

**Comment by ogencoglu** [comment]
ONNX quantization methods that don't require calibration data would be amazing.

**Comment by kshitizgupta21** [comment]
> Out-of-the-box deployment with trtllm-serve, [NVIDIA NIM](https://www.nvidia.com/en-us/ai/), and vLLM serve.

Can you elaborate on trtllm-serve? Are there any integrations planned with Triton TensorRT-LLM backend?

**Comment by wxsms** [comment]
Would like to see 4bit quantization and multi-gpu solutions on diffusion models (expecially DiT models like SD3). Thank you for the awesome work in advance!

A 4bit solution for reference: https://github.com/mit-han-lab/nunchaku

**Comment by i-riyad** [comment]
> ONNX quantization methods that don't require calibration data would be amazing.

If the users just wanna test out the latency of the model, then they can skip the calibration data and _modelopt.onnx.quantization_ will use random data to calibrate the model but the accuracy of the model will be unrealistic. Real calibration data is required by PTQ for the acceptable accuracy of the model.

**Comment by hchings** [comment]
> Would like to see 4bit quantization and multi-gpu solutions on diffusion models (expecially DiT models like SD3). Thank you for the awesome work in advance!

Hi @wxsms, thanks for the question. We have multi-GPU DiT work ongoing with the PyTorch deployment. We might include that in the early access of in-framework deployment depends on the progress. For deployment with NVIDIA TensorRT, there's ongoing work for FLUX but no timeline to be shared yet.

**Comment by ChenhanYu** [comment]
> > Out-of-the-box deployment with trtllm-serve, [NVIDIA NIM](https://www.nvidia.com/en-us/ai/), and vLLM serve.
> 
> Can you elaborate on trtllm-serve? Are there any integrations planned with Triton TensorRT-LLM backend?

TensorRT-LLM is developing a new hosting feature called `trtllm-serve` which should be similar to `vllm serve`. We are still working with TRTLLM regarding details. Regarding Triton TRTLLM, it supports the conventional TRTLLM checkpoints (which can generated by `modelopt.torch.export`).

**Comment by omrialmog** [comment]
Updated Roadmap here: #146
